aten/src/ATen/native/vulkan/api/Context.cpp

```diff
@@ -8,221 +8,31 @@ namespace vulkan {
-#ifndef DEBUG
-  static constexpr bool kEnableValidationLayers = false;
-#else
-  static constexpr bool kEnableValidationLayers = true;
-#endif
-};
-
-VKAPI_ATTR VkBool32 VKAPI_CALL debug_report_callback_fn(
-    const VkDebugReportFlagsEXT flags,
-    const VkDebugReportObjectTypeEXT /* object_type */,
-    const uint64_t /* object */,
-    const size_t /* location */,
-    const int32_t message_code,
-    const char* const layer_prefix,
-    const char* const message,
-    void* const /* user_data */) {
-  std::stringstream stream;
-  stream << layer_prefix << " " << message_code << " " << message << std::endl;
-  const std::string log = stream.str();
-
-  if (flags & VK_DEBUG_REPORT_ERROR_BIT_EXT) {
-    LOG(ERROR) << log;
-  } else if (flags & VK_DEBUG_REPORT_WARNING_BIT_EXT) {
-    LOG(WARNING) << log;
-  } else if (flags & VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT) {
-    LOG(WARNING) << "Performance:" << log;
-  } else if (flags & VK_DEBUG_REPORT_INFORMATION_BIT_EXT) {
-    LOG(INFO) << log;
-  } else if (flags & VK_DEBUG_REPORT_DEBUG_BIT_EXT) {
-    LOG(INFO) << "Debug: " << log;
-  }
-
-  return VK_FALSE;
-}
-
-VkInstance create_instance(const bool enable_validation_layers) {
-  std::vector<const char*> enabled_instance_layers;
-  std::vector<const char*> enabled_instance_extensions;
-
-  if (enable_validation_layers) {
-    uint32_t instance_layers_count = 0;
-    VK_CHECK(vkEnumerateInstanceLayerProperties(
-        &instance_layers_count, nullptr));
-
-    std::vector<VkLayerProperties> instance_layer_properties(
-        instance_layers_count);
-
-    VK_CHECK(vkEnumerateInstanceLayerProperties(
-        &instance_layers_count,
-        instance_layer_properties.data()));
-
-    constexpr const char* const requested_instance_layers[]{
-        // "VK_LAYER_LUNARG_api_dump",
-        "VK_LAYER_KHRONOS_validation",
-    };
-
-    for (const auto& requested_instance_layer : requested_instance_layers) {
-      for (const auto& layer : instance_layer_properties) {
-        if (strcmp(requested_instance_layer, layer.layerName) == 0) {
-          enabled_instance_layers.push_back(requested_instance_layer);
-          break;
-        }
-      }
-    }
-
-    uint32_t instance_extension_count = 0;
-    VK_CHECK(vkEnumerateInstanceExtensionProperties(
-        nullptr, &instance_extension_count, nullptr));
-
-    std::vector<VkExtensionProperties> instance_extension_properties(
-        instance_extension_count);
-
-    VK_CHECK(vkEnumerateInstanceExtensionProperties(
-        nullptr,
-        &instance_extension_count,
-        instance_extension_properties.data()));
-
-    constexpr const char* const requested_instance_extensions[]{
-      VK_EXT_DEBUG_REPORT_EXTENSION_NAME,
-    };
+Context* initialize() {
+  static const std::unique_ptr<Context> context([]() -> Context* {
+    try {
+      const Adapter adapter = runtime()->select([](const Adapter& adapter) {
+        // Select the first adapter.
+        return true;
```

Comment:
This lambda can later be replaced with something smarter to select the integrated GPU or the most computationally capable, etc.